parfive/tests/test_downloader.py

```diff
@@ -343,3 +348,69 @@ def test_ftp_http(tmpdir, httpserver):
     f = dl.download()
     assert len(f) == 2
     assert len(f.errors) == 4
+
+
+def test_default_user_agent(event_loop, httpserver, tmpdir):
+    tmpdir = str(tmpdir)
+    httpserver.serve_content('SIMPLE  = T',
+                             headers={'Content-Disposition': "attachment; filename=testfile.fits"})
+
+    dl = Downloader(loop=event_loop)
+    dl.enqueue_file(httpserver.url, path=Path(tmpdir), max_splits=None)
+
+    assert dl.queued_downloads == 1
+
+    dl.download()
+
+    assert 'User-Agent' in httpserver.requests[0].headers
+    assert httpserver.requests[0].headers['User-Agent'] == f"parfive/{parfive.__version__} aiohttp/{aiohttp.__version__} python/{sys.version[:5]}"
+
+
+def test_custom_user_agent(event_loop, httpserver, tmpdir):
+    tmpdir = str(tmpdir)
+    httpserver.serve_content('SIMPLE  = T',
+                             headers={'Content-Disposition': "attachment; filename=testfile.fits"})
+
+    dl = Downloader(loop=event_loop, headers={'User-Agent': 'test value 299792458'})
+    dl.enqueue_file(httpserver.url, path=Path(tmpdir), max_splits=None)
+
+    assert dl.queued_downloads == 1
+
+    dl.download()
+
+    assert 'User-Agent' in httpserver.requests[0].headers
+    assert httpserver.requests[0].headers['User-Agent'] == "test value 299792458"
+
+
+@pytest.fixture
+def create_session(loop):
+    session = None
+
+    async def maker(*args, **kwargs):
+        nonlocal session
+        session = aiohttp.ClientSession(*args, **kwargs)
+        return session
+    yield maker
+    if session is not None:
+        loop.run_until_complete(session.close())
+
+
+@pytest.fixture
+def session(create_session, loop):
+    return loop.run_until_complete(create_session())
+
+
+def test_proxy_passed_as_kwargs_to_get(session):
+    kwargs = {}
```

Comment:
This test is kinda self-fulfilling, which is a danger of mocking things.

What you should do is set some env vars at the start of the test, then mock out `ClientSession._request` like you have here, then call `parfive.Downloader.download` and then assert that the kwargs have been correctly passed to the mock, like you have here.

Hope that makes sense :smile: